Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,358 @@
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*>& textNodeList, BoundaryPointIsAtEnd isEnd)
+{
+    unsigned currentLength = 0;
+    
+    for (auto* currentNode : textNodeList) {
+        unsigned nodeEnd = currentLength + currentNode->length();
+        if (isEnd == BoundaryPointIsAtEnd::Yes)
+            nodeEnd++;
+        if (nodeEnd > index)
+            return BoundaryPoint(*currentNode, index - currentLength);
+        currentLength += currentNode->length();
+    }
+    return std::nullopt;
+}
+
+static bool indexIsWordBoundary(const String& string, unsigned index)
+{
+    UBreakIterator* it = wordBreakIterator(string);
+    return ubrk_isBoundary(it, index);
+}
+
+static bool isVisibleTextNode(const Node& node)
+{
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange& searchRange, Vector<Node*>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    String searchBuffer;
+    std::optional<BoundaryPoint> start;
+    std::optional<BoundaryPoint> end;
+    
+    if (nodes.size() < 1)
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto* node : nodes) {
+        if (node)
+            searchBufferBuilder.append(downcast<Text>(*node).data());
+    }
+    searchBuffer = searchBufferBuilder.toString();
+    
+    unsigned searchStart = 0;
+    
+    if (nodes[0] == &searchRange.startContainer())
+        searchStart = searchRange.startOffset();
+    
+    std::optional<unsigned> matchIndex;
+    
+    while (!matchIndex) {
+        // TODO: find only using the base characters i.e. also fold accents and others.
+        auto potentialIndex = searchBuffer.findIgnoringASCIICase(query, searchStart);
 
-    auto searchRange = makeRangeSelectingNodeContents(document);
+        if (potentialIndex == notFound)
+            return std::nullopt;
+        matchIndex = potentialIndex;
+        
+        start = getBoundaryPointAtIndex(*matchIndex, nodes, BoundaryPointIsAtEnd::No);
+        end = getBoundaryPointAtIndex(*matchIndex + query.length(), nodes, BoundaryPointIsAtEnd::Yes);
+        
+        if (((wordStartBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex)) || ((wordEndBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex + query.length()))) {
+            searchStart = *matchIndex + 1;
+            matchIndex = std::nullopt;
+        }
+    }
+        
+    unsigned endInset = 0;
+    if (nodes.last() == &searchRange.endContainer())
+        endInset = searchRange.endContainer().length() - searchRange.endOffset();
+    
+    if (*matchIndex + query.length() > searchBuffer.length() - endInset)
+        return std::nullopt;
+    
+    RELEASE_ASSERT_WITH_MESSAGE(start, "Scroll To Text Fragment: Start cannot be null");
```

Comment:
RELEASE_ASSERT is very aggressive; do you think it's appropriate to kill the web process just because of a text fragment lookup failure? I think probably these should be normal asserts, or should just bail from lookup, or something... not this.